controllers/metal3.io/hostfirmwarecomponents_controller.go

```diff
@@ -0,0 +1,319 @@
+	reqLogger := r.Log.WithValues("hostfirmwarecomponents", req.NamespacedName)
+	reqLogger.Info("start")
+
+	// Get the corresponding baremetalhost in this namespace, if one doesn't exist don't continue processing
+	bmh := &metal3api.BareMetalHost{}
+	err = r.Get(ctx, req.NamespacedName, bmh)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		reqLogger.Info("could not get baremetal host, not running hostfirmwarecomponents reconciler")
+		return ctrl.Result{Requeue: true, RequeueAfter: resourceNotAvailableRetryDelay}, nil
+	}
+
+	if hasDetachedAnnotation(bmh) {
+		reqLogger.Info("the host is detached, not running hostfirmwarecomponents reconciler")
+		return ctrl.Result{Requeue: true, RequeueAfter: unmanagedRetryDelay}, nil
+	}
+	// If the reconciliation is paused, requeue
+	annotations := bmh.GetAnnotations()
+
+	if _, ok := annotations[metal3api.PausedAnnotation]; ok {
+		reqLogger.Info("host is paused, no work to do")
+		return ctrl.Result{Requeue: true, RequeueAfter: subResourceNotReadyRetryDelay}, nil
+	}
+
+	// Fetch the HostFirmwareComponents
+	hfc := &metal3api.HostFirmwareComponents{}
+	info := &rhfcInfo{ctx: ctx, log: reqLogger, hfc: hfc, bmh: bmh}
+	if err = r.Get(ctx, req.NamespacedName, hfc); err != nil {
+		// The HFC resource may have been deleted
+		if k8serrors.IsNotFound(err) {
+			reqLogger.Info("hostFirmwareComponnets not found")
+			return ctrl.Result{Requeue: false}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, fmt.Errorf("could not load hostFirmwareComponents: %w", err)
+	}
+
+	// Create a provisioner to access Ironic API
+	prov, err := r.ProvisionerFactory.NewProvisioner(ctx, provisioner.BuildHostDataNoBMC(*bmh), info.publishEvent)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to create provisioner: %w", err)
+	}
+
+	ready, err := prov.TryInit()
+	if err != nil || !ready {
+		var msg string
+		if err == nil {
+			msg = "not ready"
+		} else {
+			msg = err.Error()
+		}
+		reqLogger.Info("provisioner is not ready", "Error", msg, "RequeueAfter", provisionerRetryDelay)
+		return ctrl.Result{Requeue: true, RequeueAfter: provisionerRetryDelay}, nil
+	}
+
+	newStatus, err := r.updateHostFirmware(info)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("could not update hostfirmwarecomponents: %w", err)
+	}
+
+	// Check ironic for the components information if possible
+	components, err := prov.GetFirmwareComponents()
+	info.log.Info("retrieving firmware components and saving to resource", "node", bmh.Status.Provisioning.ID)
+
+	if err != nil {
+		reqLogger.Error(err, "provisioner returns error", "RequeueAfter", provisionerRetryDelay)
+		return ctrl.Result{Requeue: true, RequeueAfter: provisionerRetryDelay}, nil
+	}
+
+	if err = r.updateHostFirmwareComponents(newStatus, components, info); err != nil {
+		return ctrl.Result{Requeue: true, RequeueAfter: provisionerRetryDelay}, nil
```

Comment:
This is probably a case for returning an error instead of just requeueing.

Reply:
ok

Reply:
Changed to 
```
return ctrl.Result{Requeue: false}, err
```